bug(workspace): renovate can't install dependencies because nrwl deps are fixed and nx-create-migrations.sh does no exist

Hi again, 

first thank you very much for your efforts in bringing all this tooling to nx. 

I finally found some time to apply your workspace tools to  a clean nx angular workspace project:
https://github.com/WolfSoko/nx-clean

Don't wonder about the missing renovate.yml and configs. The are in a different repo.  
You can find them here: 
https://github.com/WolfSoko/nx-renovate-self-hosted 

Currently there are some problems running the migration for nrwl with renovate: 

1. it looks like your nx deps are too strict. Therefore `npm ci --ignore-scripts` fails.
2. nx-create-migrations.sh is not found. I also didn't found it to be generated.  

The other errors are IMO just follow up errors.

Have a look here: 
https://github.com/WolfSoko/nx-clean/pull/38

```
File name: package.json
Command failed: npm ci --ignore-scripts
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: @nx-squeezer/workspace@1.2.0
npm ERR! Found: nx@15.6.3
npm ERR! node_modules/nx
npm ERR!   dev nx@"15.6.3" from the root project
npm ERR!   peer nx@">= 14 <= 16" from @nrwl/devkit@15.6.2
npm ERR!   node_modules/@nrwl/devkit
npm ERR!     @nrwl/devkit@"15.6.2" from @nx-squeezer/workspace@1.2.0
npm ERR!     node_modules/@nx-squeezer/workspace
npm ERR!       dev @nx-squeezer/workspace@"^1.2.0" from the root project
npm ERR!     @nrwl/devkit@"15.6.2" from ngx-deploy-npm@4.3.11
npm ERR!     node_modules/ngx-deploy-npm
npm ERR!       ngx-deploy-npm@"4.3.11" from @nx-squeezer/workspace@1.2.0
npm ERR!       node_modules/@nx-squeezer/workspace
npm ERR!         dev @nx-squeezer/workspace@"^1.2.0" from the root project
npm ERR!   10 more (@nrwl/cli, @nrwl/workspace, @nrwl/devkit, @nrwl/devkit, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer nx@"15.6.2" from @nx-squeezer/workspace@1.2.0
npm ERR! node_modules/@nx-squeezer/workspace
npm ERR!   dev @nx-squeezer/workspace@"^1.2.0" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: nx@15.6.2
npm ERR! node_modules/nx
npm ERR!   peer nx@"15.6.2" from @nx-squeezer/workspace@1.2.0
npm ERR!   node_modules/@nx-squeezer/workspace
npm ERR!     dev @nx-squeezer/workspace@"^1.2.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /home/ubuntu/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ubuntu/.npm/_logs/2023-01-29T12_54_16_041Z-debug-0.log

File name: package.json
Command failed: npm run prepare --if-present
sh: 1: husky: not found

File name: package.json
Command failed: bash nx-create-migrations.sh
bash: nx-create-migrations.sh: No such file or directory

File name: package.json
Command failed: npx --no-install nx migrate @nrwl/angular --from=@nrwl/angular@15.6.2 --to=@nrwl/angular@15.6.3
npm ERR! canceled
```

--- a/packages/workspace-e2e/tests/workspace.spec.ts
+++ b/packages/workspace-e2e/tests/workspace.spec.ts
@@ -41,6 +41,7 @@ import {
   deprecationRule,
   esLintRule,
   sonarJSRule,
+  renovateCreateMigrationsFile,
 } from '@nx-squeezer/workspace';
 
 jest.setTimeout(120_000);
@@ -222,6 +223,7 @@ describe('@nx-squeezer/workspace e2e', () => {
       expect(() => checkFilesExist(renovateCiFile)).not.toThrow();
       expect(() => checkFilesExist(renovateConfigFile)).not.toThrow();
       expect(() => checkFilesExist(renovateFile)).not.toThrow();
+      expect(() => checkFilesExist(renovateCreateMigrationsFile)).not.toThrow();
       expect(() => checkFilesExist(securityFile)).not.toThrow();
 
       renovatePresets.forEach((preset) => {

--- a/packages/workspace/project.json
+++ b/packages/workspace/project.json
@@ -61,6 +61,11 @@
         "jestConfig": "packages/workspace/jest.config.ts",
         "passWithNoTests": true,
         "codeCoverage": true
+      },
+      "configurations": {
+        "watch": {
+          "watch": true
+        }
       }
     },
     "version": {

--- a/packages/workspace/src/generators/lib/git/git-make-executable.spec.ts
+++ b/packages/workspace/src/generators/lib/git/git-make-executable.spec.ts
@@ -1,0 +1,35 @@
+import { Tree } from '@nrwl/devkit';
+import { createTree } from '@nrwl/devkit/testing';
+
+import { gitMakeExecutable } from './git-make-executable';
+import { exec } from '../exec';
+
+jest.mock('../exec');
+
+describe('@nx-squeezer/workspace gitMakeExecutable', () => {
+  let tree: Tree;
+  const file = 'file.sh';
+
+  beforeEach(() => {
+    tree = createTree();
+    jest.spyOn(console, 'log').mockImplementation(() => null);
+    jest.spyOn(console, 'error').mockImplementation(() => null);
+  });
+
+  it('should execute git command', () => {
+    (exec as jest.Mock).mockReturnValue({});
+
+    expect(() => gitMakeExecutable(tree, file)).not.toThrow();
+    expect(exec).toHaveBeenCalledWith('git', ['update-index', '--chmod=+x', file], {
+      cwd: '/virtual',
+    });
+  });
+
+  it('should throw error if exec fails', () => {
+    (exec as jest.Mock).mockReturnValue({ error: '' });
+
+    expect(() => gitMakeExecutable(tree, file)).toThrowError(
+      `Could not add execution permissions to git index for file: ${file}`
+    );
+  });
+});

--- a/packages/workspace/src/generators/lib/git/git-make-executable.ts
+++ b/packages/workspace/src/generators/lib/git/git-make-executable.ts
@@ -1,0 +1,11 @@
+import { Tree } from '@nrwl/devkit';
+
+import { exec } from '../exec';
+
+export function gitMakeExecutable(tree: Tree, file: string): void {
+  const { error } = exec('git', ['update-index', '--chmod=+x', file], { cwd: tree.root });
+
+  if (error != null) {
+    throw new Error(`Could not add execution permissions to git index for file: ${file}`);
+  }
+}

--- a/packages/workspace/src/generators/lib/git/index.ts
+++ b/packages/workspace/src/generators/lib/git/index.ts
@@ -1,1 +1,2 @@
 export * from './get-git-repo';
+export * from './git-make-executable';

--- a/packages/workspace/src/generators/renovate/files/nx-create-migrations.sh
+++ b/packages/workspace/src/generators/renovate/files/nx-create-migrations.sh
@@ -1,0 +1,4 @@
+#!/bin/bash
+
+echo '{"migrations":[]}' > migrations.json
+ls -l

--- a/packages/workspace/src/generators/renovate/generator.spec.ts
+++ b/packages/workspace/src/generators/renovate/generator.spec.ts
@@ -4,6 +4,7 @@ import { parse, stringify } from 'yaml';
 
 import { renovateGenerator } from './generator';
 import { renovateSchematic } from './generator.compat';
+import { makeMigrationsScriptExecutableTask } from './make-migrations-script-executable-task';
 import { renovateCiFile, renovateFile, renovateConfigFile, renovatePresets, renovateBranch } from './renovate';
 import { renovateConfigValidatorTask } from './renovate-config-validator-task';
 import { ciFile } from '../github-workflow';
@@ -16,6 +17,7 @@ jest.mock('../lib', () => ({
 }));
 
 jest.mock('./renovate-config-validator-task');
+jest.mock('./make-migrations-script-executable-task');
 
 describe('@nx-squeezer/workspace renovate generator', () => {
   let tree: Tree;
@@ -46,6 +48,7 @@ describe('@nx-squeezer/workspace renovate generator', () => {
     tasks?.();
 
     expect(renovateConfigValidatorTask).toHaveBeenCalled();
+    expect(makeMigrationsScriptExecutableTask).toHaveBeenCalled();
   });
 
   it('should skip execution if a Renovate CI workflow already exists', async () => {

--- a/packages/workspace/src/generators/renovate/generator.ts
+++ b/packages/workspace/src/generators/renovate/generator.ts
@@ -1,6 +1,7 @@
 import { formatFiles, generateFiles, Tree } from '@nrwl/devkit';
 import { parse, parseDocument, Scalar, stringify, YAMLSeq } from 'yaml';
 
+import { makeMigrationsScriptExecutableTask } from './make-migrations-script-executable-task';
 import { renovateCiFile, renovateBranch } from './renovate';
 import { renovateConfigValidatorTask } from './renovate-config-validator-task';
 import { RenovateGeneratorSchema } from './schema';
@@ -66,5 +67,6 @@ export async function renovateGenerator(tree: Tree, options: RenovateGeneratorSc
 
   return () => {
     renovateConfigValidatorTask(tree);
+    makeMigrationsScriptExecutableTask(tree);
   };
 }

--- a/packages/workspace/src/generators/renovate/make-migrations-script-executable-task.spec.ts
+++ b/packages/workspace/src/generators/renovate/make-migrations-script-executable-task.spec.ts
@@ -1,0 +1,33 @@
+import { Tree } from '@nrwl/devkit';
+import { createTree } from '@nrwl/devkit/testing';
+
+import { makeMigrationsScriptExecutableTask } from './make-migrations-script-executable-task';
+import { renovateCreateMigrationsFile } from './renovate';
+import { gitMakeExecutable } from '../lib';
+
+jest.mock('../lib');
+
+describe('@nx-squeezer/workspace makeMigrationsScriptExecutableTask', () => {
+  let tree: Tree;
+
+  beforeEach(() => {
+    tree = createTree();
+    jest.spyOn(console, 'log').mockImplementation(() => null);
+    jest.spyOn(console, 'error').mockImplementation(() => null);
+  });
+
+  it('should execute git command', () => {
+    (gitMakeExecutable as jest.Mock).mockReturnValue(null);
+
+    expect(makeMigrationsScriptExecutableTask(tree)).toBeTruthy();
+    expect(gitMakeExecutable).toHaveBeenCalledWith(tree, renovateCreateMigrationsFile);
+  });
+
+  it('should return false if command fails', () => {
+    (gitMakeExecutable as jest.Mock).mockImplementation(() => {
+      throw new Error(`Custom error`);
+    });
+
+    expect(makeMigrationsScriptExecutableTask(tree)).toBeFalsy();
+  });
+});

--- a/packages/workspace/src/generators/renovate/make-migrations-script-executable-task.ts
+++ b/packages/workspace/src/generators/renovate/make-migrations-script-executable-task.ts
@@ -1,0 +1,15 @@
+import { Tree } from '@nrwl/devkit';
+
+import { renovateCreateMigrationsFile } from './renovate';
+import { gitMakeExecutable } from '../lib';
+
+export function makeMigrationsScriptExecutableTask(tree: Tree): boolean {
+  try {
+    gitMakeExecutable(tree, renovateCreateMigrationsFile);
+  } catch (error) {
+    console.error(error);
+    return false;
+  }
+
+  return true;
+}

--- a/packages/workspace/src/generators/renovate/renovate.ts
+++ b/packages/workspace/src/generators/renovate/renovate.ts
@@ -1,6 +1,7 @@
 export const renovateCiFile = './.github/workflows/renovate.yml';
 export const renovateConfigFile = '.github/renovate-config.js';
 export const renovateFile = '.github/renovate.json';
+export const renovateCreateMigrationsFile = 'nx-create-migrations.sh';
 export const renovatePresets = [
   'angularWorkspace.json',
   'tooling.json',

